### PR TITLE
perf(pool): replace PoolTransferInTx.getWhere with TxPoolTransferRegistry PK lookup (#629)

### LINKED ISSUE #629: perf(pool): replace PoolTransferInTx.getWhere with TxPoolTransferRegistry lookup
## Parent PRD

Parent issue: #591 (rolls up to PRD #589)

## What to build

Eliminate the `PoolTransferInTx.getWhere({ txHash: { _eq } })` index scan in `Pool.Mint` / `Pool.Burn` handlers by adding a per-(tx, pool) registry entity. See #591 for full diagnosis and metrics.

Because the consumer (`Pool.Mint`/`Pool.Burn`) already knows the pool address (`event.srcAddress`), the registry is keyed by `${chainId}-${txHash}-${poolAddress}` — finer grain than the CL path, keeping registry rows small.

**Producer** (`storeTransferForMatching` in `src/EventHandlers/Pool/PoolTransferLogic.ts:172`):
- Upsert `TxPoolTransferRegistry` with the new `PoolTransferInTx` ID appended to `transferIds`.

**Consumer** (`getTransfersInTx` in `src/EventHandlers/Pool/PoolBurnAndMintLogic.ts:33`):
- Replace `context.PoolTransferInTx.getWhere({ txHash: { _eq } })` with `context.TxPoolTransferRegistry.get(registryId)`.
- Then `N × context.PoolTransferInTx.get(transferId)`.
- Keep the existing filter (`isMint`/`isBurn`, `logIndex < eventLogIndex`, `value > 0n`, `consumedByLogIndex` unset) and closest-preceding selection logic.

**Cleanup decision** (document in the PR):
- Today, `PoolTransferInTx` is *not* deleted on consumption — only `consumedByLogIndex` is set.
- Option (a): leave consumption as-is; registry row lives until ad-hoc GC.
- Option (b): delete `PoolTransferInTx` on consumption and prune the registry (symmetric with CL path).
- Prefer (b) unless a grep finds other readers of `PoolTransferInTx` beyond `getTransfersInTx`.

## Acceptance criteria

- [ ] New `TxPoolTransferRegistry` entity in \`schema.graphql\` (id = `{chainId}-{txHash}-{poolAddress}`, field `transferIds: [String!]!`)
- [ ] New `TxPoolTransferRegistryId` helper in \`src/Constants.ts\` with JSDoc
- [ ] `storeTransferForMatching` upserts the registry
- [ ] `getTransfersInTx` uses the registry path — no remaining `PoolTransferInTx.getWhere` calls (grep-verified)
- [ ] Cleanup strategy chosen + implemented, with a sentence in the PR description explaining why
- [ ] Existing `test/EventHandlers/Pool/PoolBurnAndMintLogic.test.ts` + `test/EventHandlers/Pool/PoolTransferLogic.test.ts` tests still pass
- [ ] New test: multi-mint or multi-burn tx → each Mint/Burn matches the correct Transfer
- [ ] New test: address(1) MINIMUM_LIQUIDITY mint edge case still resolves to the user mint (not address(1))
- [ ] `pnpm envio codegen`, `tsc --noEmit`, `pnpm qa --write`, `pnpm test` all green
- [ ] Post-merge Prometheus check: `envio_storage_load_seconds{operation="PoolTransferInTx.getWhere.txHash.eq"}` drops to ~0 on the deployed indexer

## Blocked by

None — can start immediately (parallel with the CL-path sibling).

## User stories addressed

- PRD #589 optimization tracking
- #591 acceptance criteria (PoolTransferInTx row)

## Notes

- PR base branch: \`main\`
- PR should reference #591 (close via "Contributes to #591"; #591 closes when both this and the CL-path sibling ship)

### PR DESCRIPTION
**Pull request by FonsSicca**
## Summary

- Eliminates the `PoolTransferInTx.getWhere({ txHash })` index scan in the `Pool.Mint` / `Pool.Burn` consumer (`getTransfersInTx`) by introducing a per-(tx, pool) registry entity.
- **Producer**: `storeTransferForMatching` upserts `TxPoolTransferRegistry` keyed by `{chainId}-{txHash}-{poolAddress}` with each new `PoolTransferInTx` id.
- **Consumer**: `getTransfersInTx` PK-looks up the registry then PK-`get`s each transfer by id. Existing filter (`isMint`/`isBurn`, `logIndex < eventLogIndex`, `value > 0n`, `consumedByLogIndex` unset) and closest-preceding-logIndex selection are preserved.
- **Cleanup (option b from the issue)**: on consumption, the matched `PoolTransferInTx` is deleted and its id is pruned from the registry; the row is deleted when it empties. Grep-verified that `PoolBurnAndMintLogic.ts` is the only reader of `PoolTransferInTx`, so deleting post-consume is safe.
- Grep-verified: zero remaining `PoolTransferInTx.getWhere` call sites.

Contributes to #591.

## Test plan

- [x] `pnpm envio codegen`, `tsc --noEmit`, `biome check --write`, `pnpm test` all green (1064 tests pass)
- [x] Existing `test/EventHandlers/Pool/PoolTransferLogic.test.ts` + `test/EventHandlers/Pool/PoolBurnAndMintLogic.test.ts` pass unchanged in behavior
- [x] New test: multi-mint tx (2 mints in the same (tx, pool)) — each Mint consumes its own transfer, registry pruned and ultimately deleted
- [x] New test: address(1) MINIMUM_LIQUIDITY mint edge case — still resolves to the user mint; consumed (deleted) transfer is the user one, not address(1)
- [ ] **Post-merge Prometheus check**: `envio_storage_load_seconds{operation="PoolTransferInTx.getWhere.txHash.eq"}` drops to ~0 (tracked via #630)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Refactor**
  * Introduced a per-transaction/per-pool registry to track ordered pool transfer entries, improving matching and reducing broad lookups.
  * Consumption now removes matched transfer entries (no longer uses a persisted "consumed" marker).

* **Tests**
  * Updated and expanded tests to cover registry-backed matching, multi-mint/burn ordering, and deletion/pruning behavior.

### CODE REVIEW

**Comment by FonsSicca** [comment]
@coderabbitai review

**Comment by FonsSicca** [comment]
@coderabbitai review

**Review by WooSungD — APPROVED** [review]

File: test/EventHandlers/Pool/PoolBurnAndMintLogic.test.ts [hunk 1]
```diff
@@ -49,25 +51,68 @@ describe("PoolBurnAndMintLogic", () => {
...
+          const idx = mockRegistries.findIndex((r) => r.id === entity.id);
+          if (idx >= 0) mockRegistries[idx] = entity;
+          else mockRegistries.push(entity);
+        }),
+        deleteUnsafe: vi.fn((id: string) => {
+          const idx = mockRegistries.findIndex((r) => r.id === id);
+          if (idx >= 0) mockRegistries.splice(idx, 1);
         }),
-        set: vi.fn(),
       },
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply test/EventHandlers/Pool/PoolBurnAndMintLogic.test.ts:121]: > a one-line comment upgrade or a `seeded` flag (instead of the length comparison) would make the harness harder to misuse in the future.

Addressed in e0daafb. Replaced the `mockRegistries.length === 0 && mockPoolTransferInTx.length > 0` guard with an explicit `registrySeeded` flag that's set true by `seedRegistriesFromTransfers`. A test that calls `deleteUnsafe` on a registry row can no longer accidentally resurrect it from stale `PoolTransferInTx` rows on the next `get`, so the harness now matches production's no-resurrection semantics.

**Comment by FonsSicca** [comment]
Addressed the 4 outside-diff findings in 2f753fc (plus rebased onto main to pick up #631; force-pushed):

> **`src/EventHandlers/Pool/PoolBurnAndMintLogic.ts` (189-199): Stale comment: no more `txHash` index scan.**

Comment updated to describe the new PK-read flow: `TxPoolTransferRegistry` PK lookup, then PK `get`s per transfer id.

> **`src/EventHandlers/Pool/PoolBurnAndMintLogic.ts` (101-111): Dead filter branch: `consumedByLogIndex` is never set anymore.**

Followed through on the cleanup. Consumed transfers are now `deleteUnsafe`d (not marked), so the field is genuinely write-never for `PoolTransferInTx`. Removed:
- The filter clause in `getPrecedingTransfers`
- The `consumedByLogIndex: undefined` initializer in `storeTransferForMatching`
- The `consumedByLogIndex` field from `PoolTransferInTx` in `schema.graphql`
- Two now-meaningless tests (`should exclude already consumed transfers`, `should handle null consumedByLogIndex`) and the `consumedByLogIndex` param from the `createMockTransfer` helper

ALM's separate `ALMLPWrapperTransferInTx.consumedByLogIndex` still has live writers (`LPWrapperLogic.ts:259`), so it's untouched.

> **`test/EventHandlers/Pool/PoolTransferLogic.test.ts` (308-385): Await the now-async `storeTransferForMatching` and assert the registry side effect.**

All three tests now `await` `storeTransferForMatching`. The mint/burn cases assert `TxPoolTransferRegistry.set` was called with the expected `{ id, transferIds: [transferId] }`; the "should not store regular transfers" case asserts `TxPoolTransferRegistry.set` was **not** called.

> **`src/EventHandlers/Pool/PoolTransferLogic.ts` (160-226): Update JSDoc to document the new registry side effect, ordering guarantee, and async contract.**

Expanded `storeTransferForMatching` JSDoc with a side-effects paragraph covering (1) the `PoolTransferInTx` upsert, (2) the `TxPoolTransferRegistry` upsert with insertion-order append (consumer relies on this for closest-preceding selection), and the last-writer-wins semantics on the registry row. Added the `@returns` line.

Full suite stays green: `tsc --noEmit` clean, 1063 tests pass, lint clean on changed files.

**Comment by FonsSicca** [comment]
@coderabbitai review

**Comment by FonsSicca** [comment]
@coderabbitai review

**Comment by FonsSicca** [comment]
@coderabbitai review
